### PR TITLE
Optimize getSingleValueBlock in Map and Array

### PR DESCRIPTION
**Pull request by arunthirupathi**
getSingleValueBlock is used by nested loop joins in broadcast queries.
getSingleValueBlock unconditionally clones even when the underlying
block has only one element. Map and Arrays are used in aggregation
and broad casted. This path is always taken or never taken so the cost
of branching should be minimal.

Test plan - 
Added new tests.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-common/src/main/java/com/facebook/presto/common/block/ArrayBlock.java [hunk 1]
```diff
@@ -208,4 +208,24 @@ public Block getLoadedBlock()
                 offsets,
                 loadedValuesBlock);
     }
+
+    private boolean isSinglePositionBlock(int position)
+    {
+        // Verify the array has only one element
+        if (position == 0 && positionCount == 1) {
+            // verify offsets array has only one element.
+            return offsets.length == 2;
```
Comment (sdruzkin) [inline presto-common/src/main/java/com/facebook/presto/common/block/ArrayBlock.java:217]: `offsets.length == 2` can be wrapped into the `if` condition to return true here.

File: presto-common/src/main/java/com/facebook/presto/common/block/Block.java [hunk 2]
```diff
@@ -164,7 +164,7 @@ default int compareTo(int leftPosition, int leftOffset, int leftLength, Block ri
 
     /**
      * Gets the value at the specified position as a single element block.  The method
-     * must copy the data into a new block.
+     * must copy the data into a new block, if there are more elements in the block.
```
Comment (sdruzkin) [inline presto-common/src/main/java/com/facebook/presto/common/block/Block.java:167]: I'd rephrase it with "more than one element", something like this:
`if there are more elements in the block` -> `if there are more than one element in the block`

File: presto-main/src/test/java/com/facebook/presto/block/TestArrayBlock.java [hunk 3]
```diff
@@ -110,6 +112,58 @@ public void testWithArrayBlock()
         assertBlockFilteredPositions(expectedValuesWithNull, blockBuilderWithNull.build(), () -> blockBuilder.newBlockBuilderLike(null), 2, 3, 4, 9, 13, 14);
     }
 
+    @Test
+    public void testSingleValueBlock()
+    {
+        // 1 entry array.
+        long[][] values = createTestArray(50);
+        // ArrayBlock's hold onto the offsets from builder. MapBlock copies the required portion of the offsets into a new array.
```
Comment (sdruzkin) [inline presto-main/src/test/java/com/facebook/presto/block/TestArrayBlock.java:120]: 1. Rewrap this comment into two lines. 
2. `ArrayBlock's hold` looks like a possessive voice, see if it would look better to you if change it to 'ArrayBlock holds` similar to MapBlock in the next sentence.

File: presto-main/src/test/java/com/facebook/presto/block/TestArrayBlock.java [hunk 4]
```diff
@@ -110,6 +112,58 @@ public void testWithArrayBlock()
...
+
+        assertSame(arrayBlock, arrayBlock.getSingleValueBlock(0));
+        assertNotSame(arrayBlockBuilder, arrayBlockBuilder.getSingleValueBlock(0));
+
+        // 2 entries array.
+        values = createTestArray(50, 50);
+        arrayBlockBuilder = createBlockBuilderWithValues(values);
+        arrayBlock = arrayBlockBuilder.build();
+        Block firstElement = arrayBlock.getRegion(0, 1);
+        assertNotSame(firstElement, firstElement.getSingleValueBlock(0));
```
Comment (sdruzkin) [inline presto-main/src/test/java/com/facebook/presto/block/TestArrayBlock.java:134]: Add asserts for all subsequent arrayBlockBuilders too?

File: presto-main/src/test/java/com/facebook/presto/block/TestArrayBlock.java [hunk 5]
```diff
@@ -110,6 +112,58 @@ public void testWithArrayBlock()
...
+        assertSame(arrayBlock, arrayBlock.getSingleValueBlock(0));
+
+        // Test with 2 null elements.
+        values = new long[][] {null, null};
+        arrayBlockBuilder = createBlockBuilderWithValues(values);
+        arrayBlock = arrayBlockBuilder.build();
+        assertNotSame(arrayBlock, arrayBlock.getSingleValueBlock(0));
+    }
+
+    private long[][] createTestArray(int... entryCounts)
```
Comment (sdruzkin) [inline presto-main/src/test/java/com/facebook/presto/block/TestArrayBlock.java:153]: static

File: presto-main/src/test/java/com/facebook/presto/block/TestArrayBlock.java [hunk 6]
```diff
@@ -110,6 +112,60 @@ public void testWithArrayBlock()
         assertBlockFilteredPositions(expectedValuesWithNull, blockBuilderWithNull.build(), () -> blockBuilder.newBlockBuilderLike(null), 2, 3, 4, 9, 13, 14);
     }
 
+    @Test
+    public void testSingleValueBlock()
+    {
+        // 1 entry array.
+        long[][] values = createTestArray(50);
+        // ArrayBlock holds onto the offsets from builder. Offsets can contain additional elements.
+        // MapBlock copies the required portion of the offsets into a new array.
```
Comment (rschlussel) [inline presto-main/src/test/java/com/facebook/presto/block/TestArrayBlock.java:121]: I think you mean ArrayBlock here

**Review by rschlussel — APPROVED** [review]

File: presto-main/src/test/java/com/facebook/presto/block/TestArrayBlock.java (hunk 6 above)
Comment (arunthirupathi) [reply presto-main/src/test/java/com/facebook/presto/block/TestArrayBlock.java:121]: I will remove the comment. I tried using the createBlockBuilderWithValues first, but it sets expected entries to 100 and singleValueBlock returned a different block. I tried capturing that in the comment, which is more confusing.

**Review by sdruzkin — APPROVED** [review]

**Comment by arunthirupathi** [comment]
This commit is cherry-picked into https://github.com/prestodb/presto/pull/16346 for merging.
